backends/dpdk/dpdkAsmOpt.cpp

```diff
@@ -194,4 +208,65 @@ const IR::Node* RemoveUnusedMetadataFields::preorder(IR::DpdkAsmProgram *p) {
     return p;
 }
 
+bool ValidateTableKeys::isMetadataStruct(const IR::Type_Struct *st) {
+    for (auto anno : st->annotations->annotations) {
+        if (anno->name == "__metadata__") {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool ValidateTableKeys::preorder(const IR::DpdkAsmProgram *p) {
+    const IR::DpdkStructType *metaStruct;
+    for (auto st : p->structType) {
+        if (isMetadataStruct(st)) {
+            metaStruct = st;
+            break;
+        }
+    }
+    for (auto tbl : p->tables) {
+        int min, max, size_max_field = 0;
+        auto keys = tbl->match_keys;
+        if (!keys || keys->keyElements.size() == 0) {
+            return false;
+        }
+        min = max = -1;
+        for (auto key : keys->keyElements) {
+            BUG_CHECK(key->expression->is<IR::Member>(), "Table keys must be a structure field. "
+                                                          "%1% is not a structure field", key);
+            auto keyMem = key->expression->to<IR::Member>();
+            auto type = keyMem->expr->type;
+            if (type->is<IR::Type_Struct>() && isMetadataStruct(type->to<IR::Type_Struct>())) {
+                auto offset = metaStruct->getFieldBitOffset(keyMem->member.name);
+                if (min == -1 || min > offset)
+                    min = offset;
+                if (max == -1 || max < offset) {
+                    max = offset;
+                    auto field_type = key->expression->type;
+                    if (auto t = field_type->to<IR::Type_Bits>()) {
```

Comment:
you may want to make this a utility function

Reply:
Done.